### ISSUE TITLE
[Ecosystem] AReaL

### ISSUE THREAD

**Issue by garrett4wade** [pytorch-fdn/ecosystem#65]
### Contact emails

fuwth17@gmail.com

### Project summary

AReaL is a scalable, PyTorch-native reinforcement learning (RL) infrastructure designed to bridge foundation model training with modern agent-based applications.

### Project description

AReaL is a modular RL infrastructure that connects foundation model training
with downstream agent applications. Built on a fully asynchronous RL training paradigm,
AReaL provides:

- **Modular Architecture**: Decomposes RL into independent, composable services
  (agent, inference, training, and weight updates) with well-defined interfaces.
  This design enables flexible scaling, fault tolerance, and independent optimization
  of system components.

- **Plug-and-Play RL as a Service**: Standardized data and execution interfaces allow
  users to apply RL to existing systems, including black-box or online systems, with
  minimal integration overhead and no intrusive code changes.

- **Ecosystem Compatibility**: A decoupled system design enables seamless integration
  with diverse training and inference backends, including vLLM, SGLang, Megatron,
  and custom PyTorch-native 5D parallel engines.

### Are there any other projects in the PyTorch Ecosystem similar to yours?  If, yes, what are they?

Yes, including torchrl, torchforge, and verl.

### Project repo URL

https://github.com/inclusionAI/AReaL

### Additional repos in scope of the application

N/A

### Project license

Apache-2.0

### GitHub handles of the project maintainer(s)

@garrett4wade, @rchardx, @nuzant, @fishcrap

### Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

AReaL Team, Ant Group (https://www.antgroup.com) and IIIS, Tsinghua University (https://iiis.tsinghua.edu.cn/en/)

### Website URL

https://inclusionai.github.io/AReaL/

### Documentation

- https://inclusionai.github.io/AReaL/en/ (English)
- https://inclusionai.github.io/AReaL/zh/ (Chinese)

### How do you build and test the project today (continuous integration)?  Please describe.

AReaL uses GitHub Actions with a multi-tier CI pipeline:

**1. Pre-commit Checks** (`pre-commit.yml`) — Runs on every PR:
- Formatting, lisense remark, CLI documentation, commit convention

**2. Package Installation Tests** (`install-test.yml`) — Runs on PRs that modify
`pyproject.toml`, `uv.lock`, or `areal/`:
- Cross-platform installation (Docker Image, Ubuntu CUDA, and macOS, Python 3.12)

**3. GPU Unit & Integration Tests** (`test-areal.yml`) — Runs on PRs labeled
`safe-to-test`:
- Run tests (`./tests`) with 2× NVIDIA A100 GPUs (40GB) on GCP with the latest image
- Three test suites: unit tests (`tests/test_*.py`), SFT integration (`tests/sft/`),
  and GRPO integration (`tests/grpo/`)

**4. Docker Image Build & Promotion** (`build-docker-image.yml`) — Manual trigger:
- Builds SGLang and vLLM Docker image variants (probably with different PyTorch versions)
- Runs the full test suite on newly built images

**5. Documentation Deployment** (`deploy-docs.yml`) — Runs on push to `main`:
- Builds bilingual Jupyter Book documentation (English and Chinese)
- Deploys to GitHub Pages

### Version of PyTorch

AReaL is built and tested with:

- **PyTorch 2.9.1+cu129** - Default when using the SGLang inference backend
- **PyTorch 2.10.0+cu129** - When using the vLLM inference backend

The minimum supported version is **PyTorch ≥ 2.9.1**, as required by the inference backend.

### Components of PyTorch

**Dependencies:**
- transformers, Megatron, SGLang, vLLM, etc, all built on PyTorch

**Data Flow:**
- `torch.Tensor` as the default representation for data such as token IDs, attention masks,
  and log probabilities
- Wrapped as `areal.rpc.infra.rtensor.RTensor` for distributed storage and transfer

**Core Distributed Training Infrastructure:**
- `torch.distributed` — Process groups and collective operations (all_reduce, broadcast,
  all_gather, reduce_scatter, all_to_all)
- `torch.distributed.device_mesh.DeviceMesh` — N-dimensional device mesh for composing
  parallelism strategies (data, tensor, context, expert, and pipeline parallelism)

**Customized 5D Parallelism Engine (Archon, inspired by `torchtitan`):**
- `torch.distributed.fsdp` (FSDP2) — Fully Sharded Data Parallel with per-parameter sharding,
  `MixedPrecisionPolicy`, and `CPUOffloadPolicy`
- `torch.distributed.tensor.DTensor` — Distributed tensor abstraction with placement types
  (`Replicate`, `Shard`, `_StridedShard`)
- `torch.distributed.tensor.parallel` — `parallelize_module()` and parallelization strategies
- `torch.distributed.pipelining` — Pipeline stages and scheduling (GPipe, 1F1B, interleaved 1F1B)
- `torch.distributed.checkpoint` (DCP) — APIs for saving/loading sharded model states
- `torch.compile` (Inductor backend) — Graph-level optimization with dynamic shape support

**Utilities from the Ecosystem:**
- `torch.profiler` — Performance profiling and Chrome trace export
- `torchdata.stateful_dataloader.StatefulDataLoader` — Resumable data loading
- `torchao` — FP8 quantization support

### How long do you expect to maintain the project?

We plan to maintain AReaL until AGI benefits everyone in the real life.

### Additional information

N/A

**Comment by HwVanICI** [comment]
We would like to express our support for AReaL’s application to join the PyTorch Ecosystem.

Our engineering team at Huawei has adopted AReaL internally and has also contributed to the project over time. Based on our experience as both users and contributors, we believe AReaL is a strong fit for the ecosystem.

Its PyTorch-native design and modular architecture make it practical for modern RL workloads, agent-oriented applications, and distributed training scenarios. We have also had a positive experience collaborating with the project and its maintainers, and have found it open to external feedback and contribution.

We are happy to support this application.

**Comment by michellearoth** [comment]
Accepted! 

<img width="1223" height="1967" alt="Image" src="https://github.com/user-attachments/assets/99fbb050-23cb-42d1-9dae-577abebc7b52" />
